File: src/main/java/com/couchbase/cblite/CBLManager.java
package com.couchbase.cblite;

import com.couchbase.cblite.auth.CBLAuthorizer;
import com.couchbase.cblite.auth.CBLFacebookAuthorizer;
import com.couchbase.cblite.auth.CBLPersonaAuthorizer;
import com.couchbase.cblite.replicator.CBLPusher;
import com.couchbase.cblite.replicator.CBLReplicator;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

public enum CBLManager {

    INSTANCE;

    private CBLServer server;

    public CBLServer getServer() {
        return server;
    }

    public void setServer(CBLServer server) {
        this.server = server;
    }


    private Map<String, Object> parseSourceOrTarget(Map<String,Object> properties, String key) {
        Map<String, Object> result = new HashMap<String, Object>();

        Object value = properties.get(key);

        if (value instanceof String) {
            result.put("url", (String)value);
        }
        else if (value instanceof Map) {
            result = (Map<String, Object>) value;
        }
        return result;

    }

    public CBLReplicator getReplicator(Map<String,Object> properties) throws CBLiteException {

        CBLAuthorizer authorizer = null;
        CBLReplicator repl = null;
        URL remote = null;


        Map<String, Object> remoteMap;

        Map<String, Object> sourceMap = parseSourceOrTarget(properties, "source");
        Map<String, Object> targetMap = parseSourceOrTarget(properties, "target");

        String source = (String)sourceMap.get("url");
        String target = (String)targetMap.get("url");

        Boolean createTargetBoolean = (Boolean)properties.get("create_target");
        boolean createTarget = (createTargetBoolean != null && createTargetBoolean.booleanValue());

        Boolean continuousBoolean = (Boolean)properties.get("continuous");
        boolean continuous = (continuousBoolean != null && continuousBoolean.booleanValue());

        Boolean cancelBoolean = (Boolean)properties.get("cancel");
        boolean cancel = (cancelBoolean != null && cancelBoolean.booleanValue());

        // Map the 'source' and 'target' JSON params to a local database and remote URL:
        if(source == null || target == null) {
            throw new CBLiteException("source and target are both null", new CBLStatus(CBLStatus.BAD_REQUEST));
        }

        boolean push = false;

        CBLDatabase db = getServer().getExistingDatabaseNamed(source);
        String remoteStr = null;
        if(db != null) {
            remoteStr = target;
            push = true;
            remoteMap = targetMap;
        } else {
            remoteStr = source;
            if(createTarget && !cancel) {
                db = getServer().getDatabaseNamed(target);
                if(!db.open()) {
                    throw new CBLiteException("cannot open database: " + db, new CBLStatus(CBLStatus.INTERNAL_SERVER_ERROR));
                }
            } else {
                db = getServer().getExistingDatabaseNamed(target);
            }
            if(db == null) {
                throw new CBLiteException("database is null", new CBLStatus(CBLStatus.NOT_FOUND));
            }
            remoteMap = sourceMap;
        }


        Map<String, Object> authMap = (Map<String, Object>) remoteMap.get("auth");
        if (authMap != null) {

            Map<String, Object> persona = (Map<String, Object>) authMap.get("persona");
            if (persona != null) {
                String email = (String) persona.get("email");
                authorizer = new CBLPersonaAuthorizer(email);
            }
            Map<String, Object> facebook = (Map<String, Object>) authMap.get("facebook");
            if (facebook != null) {
                String email = (String) facebook.get("email");
                authorizer = new CBLFacebookAuthorizer(email);
            }

        }

        try {
            remote = new URL(remoteStr);
        } catch (MalformedURLException e) {
            throw new CBLiteException("malformed remote url: " + remoteStr, new CBLStatus(CBLStatus.BAD_REQUEST));
        }
        if(remote == null || !remote.getProtocol().startsWith("http")) {
            throw new CBLiteException("remote URL is null or non-http: " + remoteStr, new CBLStatus(CBLStatus.BAD_REQUEST));
        }


        if(!cancel) {
            repl = db.getReplicator(remote, getServer().getDefaultHttpClientFactory(), push, continuous, getServer().getWorkExecutor());
            if(repl == null) {
                throw new CBLiteException("unable to create replicator with remote: " + remote, new CBLStatus(CBLStatus.INTERNAL_SERVER_ERROR));
            }

            if (authorizer != null) {
                repl.setAuthorizer(authorizer);
            }

            String filterName = (String)properties.get("filter");
            if(filterName != null) {
                repl.setFilterName(filterName);
                Map<String,Object> filterParams = (Map<String,Object>)properties.get("query_params");
                if(filterParams != null) {
                    repl.setFilterParams(filterParams);
                }
            }

            if(push) {
                ((CBLPusher)repl).setCreateTarget(createTarget);
            }


        } else {
            // Cancel replication:
            repl = db.getActiveReplicator(remote, push);
            if(repl == null) {
                throw new CBLiteException("unable to lookup replicator with remote: " + remote, new CBLStatus(CBLStatus.NOT_FOUND));
            }
        }

        return repl;
    }

}

File: Couchbase.Lite/Couchbase.Lite/CBLManager.cs
/**
 * Couchbase Lite for .NET
 *
 * Original iOS version by Jens Alfke
 * Android Port by Marty Schoch, Traun Leyden
 * C# Port by Zack Gramana
 *
 * Copyright (c) 2012, 2013 Couchbase, Inc. All rights reserved.
 * Portions (c) 2013 Xamarin, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

using System;
using System.Collections;
using System.Collections.Generic;
using Couchbase;
using Couchbase.Auth;
using Couchbase.Replicator;
using Sharpen;

namespace Couchbase
{
	public class CBLManager
	{
		private CBLServer server;

		//INSTANCE;
		public virtual CBLServer GetServer()
		{
			return server;
		}

		public virtual void SetServer(CBLServer server)
		{
			this.server = server;
		}

		private IDictionary<string, object> ParseSourceOrTarget(IDictionary<string, object
			> properties, string key)
		{
			IDictionary<string, object> result = new Dictionary<string, object>();
			object value = properties.Get(key);
			if (value is string)
			{
				result.Put("url", (string)value);
			}
			else
			{
				if (value is IDictionary)
				{
					result = (IDictionary<string, object>)value;
				}
			}
			return result;
		}

		/// <exception cref="Couchbase.CBLiteException"></exception>
		public virtual CBLReplicator GetReplicator(IDictionary<string, object> properties
			)
		{
			CBLAuthorizer authorizer = null;
			CBLReplicator repl = null;
			Uri remote = null;
			IDictionary<string, object> remoteMap;
			IDictionary<string, object> sourceMap = ParseSourceOrTarget(properties, "source");
			IDictionary<string, object> targetMap = ParseSourceOrTarget(properties, "target");
			string source = (string)sourceMap.Get("url");
			string target = (string)targetMap.Get("url");
			bool createTargetBoolean = (bool)properties.Get("create_target");
			bool createTarget = (createTargetBoolean != null && createTargetBoolean);
			bool continuousBoolean = (bool)properties.Get("continuous");
			bool continuous = (continuousBoolean != null && continuousBoolean);
			bool cancelBoolean = (bool)properties.Get("cancel");
			bool cancel = (cancelBoolean != null && cancelBoolean);
			// Map the 'source' and 'target' JSON params to a local database and remote URL:
			if (source == null || target == null)
			{
				throw new CBLiteException("source and target are both null", new CBLStatus(CBLStatus
					.BadRequest));
			}
			bool push = false;
			CBLDatabase db = GetServer().GetExistingDatabaseNamed(source);
			string remoteStr = null;
			if (db != null)
			{
				remoteStr = target;
				push = true;
				remoteMap = targetMap;
			}
			else
			{
				remoteStr = source;
				if (createTarget && !cancel)
				{
					db = GetServer().GetDatabaseNamed(target);
					if (!db.Open())
					{
						throw new CBLiteException("cannot open database: " + db, new CBLStatus(CBLStatus.
							InternalServerError));
					}
				}
				else
				{
					db = GetServer().GetExistingDatabaseNamed(target);
				}
				if (db == null)
				{
					throw new CBLiteException("database is null", new CBLStatus(CBLStatus.NotFound));
				}
				remoteMap = sourceMap;
			}
			IDictionary<string, object> authMap = (IDictionary<string, object>)remoteMap.Get(
				"auth");
			if (authMap != null)
			{
				IDictionary<string, object> persona = (IDictionary<string, object>)authMap.Get("persona"
					);
				if (persona != null)
				{
					string email = (string)persona.Get("email");
					authorizer = new CBLPersonaAuthorizer(email);
				}
				IDictionary<string, object> facebook = (IDictionary<string, object>)authMap.Get("facebook"
					);
				if (facebook != null)
				{
					string email = (string)facebook.Get("email");
					authorizer = new CBLFacebookAuthorizer(email);
				}
			}
			try
			{
				remote = new Uri(remoteStr);
			}
			catch (UriFormatException)
			{
				throw new CBLiteException("malformed remote url: " + remoteStr, new CBLStatus(CBLStatus
					.BadRequest));
			}
			if (remote == null || !remote.Scheme.StartsWith("http"))
			{
				throw new CBLiteException("remote URL is null or non-http: " + remoteStr, new CBLStatus
					(CBLStatus.BadRequest));
			}
			if (!cancel)
			{
				repl = db.GetReplicator(remote, GetServer().GetDefaultHttpClientFactory(), push, 
					continuous, GetServer().GetWorkExecutor());
				if (repl == null)
				{
					throw new CBLiteException("unable to create replicator with remote: " + remote, new 
						CBLStatus(CBLStatus.InternalServerError));
				}
				if (authorizer != null)
				{
					repl.SetAuthorizer(authorizer);
				}
				string filterName = (string)properties.Get("filter");
				if (filterName != null)
				{
					repl.SetFilterName(filterName);
					IDictionary<string, object> filterParams = (IDictionary<string, object>)properties
						.Get("query_params");
					if (filterParams != null)
					{
						repl.SetFilterParams(filterParams);
					}
				}
				if (push)
				{
					((CBLPusher)repl).SetCreateTarget(createTarget);
				}
			}
			else
			{
				// Cancel replication:
				repl = db.GetActiveReplicator(remote, push);
				if (repl == null)
				{
					throw new CBLiteException("unable to lookup replicator with remote: " + remote, new 
						CBLStatus(CBLStatus.NotFound));
				}
			}
			return repl;
		}
	}
}
